### DISCUSSION TITLE
Install `buildx`

### DISCUSSION THREAD

**Discussion by halostatue** [abiosoft/colima#273] [Q&A] [answered]
It doesn’t matter much from my perspective, because I’m using `buildkit` (via `DOCKER_BUILDKIT=1`), but I’m curious how one would install `buildx` on macOS for use with Colima, as `brew install docker` does not seem to include it. I can obviously install Docker desktop, but that sort of defeats the purpose.

Those of you using it, how have you installed it? Manual plug-in installation?

**Comment by abiosoft** [comment, marked as answer]
```sh
ARCH=amd64 # change to 'arm64' for m1
VERSION=v0.8.2
curl -LO https://github.com/docker/buildx/releases/download/${VERSION}/buildx-${VERSION}.darwin-${ARCH}
mkdir -p ~/.docker/cli-plugins
mv buildx-${VERSION}.darwin-${ARCH} ~/.docker/cli-plugins/docker-buildx
chmod +x ~/.docker/cli-plugins/docker-buildx
docker buildx version # verify installation
```

**Reply by jacobibanez** [reply]
Hi, thanks for this answer. I have tried these steps, though, and my terminal still says buildx is not a docker command :( I'm using NOT an m1 mac, with colima instead of docker desktop

**Reply by asciidiego** [reply]
Warning: Remember to change the `VERSION` shell variable to whatever `buildx` version you want to install.

**Reply by theBeginner86** [reply]
Thanks this was helpful!

**Reply by syedmohsinbukhari** [reply]
thanks! this worked in April 2025!

**Reply by mansoorTW** [reply]
Thanks :)

**Reply by bekzodnj** [reply]
Thanks! I was able to install buildx v.0.24.0 as of June 2025.

**Reply by ishaanbahal** [reply]
Works for buildx v0.25.0 Jun 25, can add this to the script for latest buildx release
```bash
VERSION=$(curl https://api.github.com/repos/docker/buildx/releases/latest -s | jq .name -r)
```

**Reply by arunlogo1** [reply]
When install a docker desktop, buildx comes as the package. If you install a docker cli as a standalone part, we need to install buildx as binary in plugin

**Reply by mrs83** [reply]
this worked in Sept 2025!

**Reply by sarahsturgeon** [reply]
combined and modified the scripts here to make something that should work for any OSX device _(probably)_.

installs the latest version for either architecture (`arm64` or `x86_64`/`amd64`)

certainly, it worked for my M4 in Oct 2025:
```bash
echo "Installing buildx plugin for OSX" && \
    ARCH=$(uname -m) && \
    ([[ "$ARCH" == "x86_64" ]] && ARCH="amd64" || true) && \
    VERSION=$(curl -s https://api.github.com/repos/docker/buildx/releases/latest | jq -r .name) && \
    FILENAME="buildx-${VERSION}.darwin-${ARCH}" && \
    echo "Downloading Docker Buildx ${VERSION} for ${ARCH}..." && \
    curl --fail --location --output "$FILENAME" "https://github.com/docker/buildx/releases/download/${VERSION}/${FILENAME}" && \
    mkdir -p ~/.docker/cli-plugins && \
    mv -v "$FILENAME" ~/.docker/cli-plugins/docker-buildx && \
    chmod +x ~/.docker/cli-plugins/docker-buildx && \
    echo "Successfully installed docker-buildx." && \
    docker buildx version
```

**Reply by axel-eck** [reply]
@sarahsturgeon scripts worked like a charm. Thanks

**Reply by sarahsturgeon** [reply]
👍 

After writing that, I actually decided I preferred the other solution: Installing with Homebrew and then adding the homebrew `cli-plugins` dir as a plugin dir in docker.

This way, `buildx` will be updated any time I do my normal `brew upgrade` without having to re-run a script or manually download the binary again.

```
brew install docker-buildx;
```

then add to `~/.docker/config.json`:
```json
"cliPluginsExtraDirs": [
    "/opt/homebrew/lib/docker/cli-plugins"
]
```

I also restarted colima but didn't check if it was necessary:
```
colima restart
```

_(linking the bin would also work, but this made more sense to me)_

The only _potential_ downside is if you just _needed_ some new feature/functionality from `buildx` that hadn't yet made it to the formula (no idea how realistic this is)

**Reply by flipcc** [reply]
Terrific response. Love the "brew-way"!

**Reply by aristocrates** [reply]
> I also restarted colima but didn't check if it was necessary:

I successfully followed those steps and I did **not** need to run `colima restart`

**Comment by fullheart** [comment]
Hi together,

Thank you for this solution!

I created for my this **installer Script** to install with GitHub CLI the latest `buildx` command on my Mac with M1. This script install always the latest version of the command:

https://github.com/fullheart/my-dev-env/blob/main/osx/docker/colima/install_latest_buildx.sh

**Comment by amitds1997** [comment]
For me, this worked:
```bash
brew install docker-buildx
# Follow the caveats mentioned in the install instructions:
# mkdir -p ~/.docker/cli-plugins
# ln -sfn $(which docker-buildx) ~/.docker/cli-plugins/docker-buildx
docker buildx install # To keep using the `docker build` install but with buildkit: https://docs.docker.com/engine/reference/commandline/buildx_install/
```
Edit 1: Updated specific location in `# ln -sfn /usr/local/opt/docker-buildx/bin/docker-buildx ~/.docker/cli-plugins/docker-buildx` with `$(which docker-buildx)`.

**Reply by halostatue** [reply]
This is a simpler answer if using Homebrew.

**Reply by terracota-p** [reply]
That worked for me on m1, thanks.

**Reply by peraronsson** [reply]
Following these instructions, I first got `docker: 'buildx' is not a docker command.` when running `docker buildx install`.
A small modification was required on my M1; the symlink had to be changed to 
 `ln -sfn /opt/homebrew/bin/docker-buildx ~/.docker/cli-plugins/docker-buildx`

**Reply by halostatue** [reply]
This could be generalized to `ln -sfn $(which docker-buildx) ~/.docker/cli-plugins`.

**Reply by connorads** [reply]
> Following these instructions, I first got `docker: 'buildx' is not a docker command.` when running `docker buildx install`. A small modification was required on my M1; the symlink had to be changed to `ln -sfn /opt/homebrew/bin/docker-buildx ~/.docker/cli-plugins/docker-buildx`

Thanks. I had to adjust it slightly to get it to work.
```sh
mkdir ~/.docker/cli-plugins
ln -sfn $(which docker-buildx) ~/.docker/cli-plugins/docker-buildx
```

**Reply by nprokopic** [reply]
June 2025 update 😃 after installing buildx with `brew install docker-buildx` the additional step I did (that Homebrew itself suggests) was to add the following config to `~/.docker/config.json`:

```JSON
{
    "cliPluginsExtraDirs": [
        "/opt/homebrew/lib/docker/cli-plugins"
    ]
}
```

**Reply by RobertaMurniece** [reply]
This worked perfectly on m1! Sep 2025

**Comment by arthurspa** [comment]
I addition to @amitds1997 answer, I had to set `DOCKER_HOST` to point it to Colima docker socket. E.g. in my .zshrc I added `export DOCKER_HOST="unix://$HOME/.colima/docker.sock"`. Otherwise I'd get this error:

```docker buildx ls                                                                                                                                                                                       ─╯
NAME/NODE DRIVER/ENDPOINT STATUS  BUILDKIT PLATFORMS
colima *  docker                           
  colima  colima          running 20.10.20 linux/arm64, linux/amd64
default                   error            

Cannot load builder default: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?```

Pasted here in case someone else run in the same problem. I'm using Mac M1.

**Reply by halostatue** [reply]
This would be better to do as:

`export DOCKER_HOST="$(docker context inspect --format '{{ .Endpoints.docker.Host }}')"`

It’s worth noting that if you have *multiple* Colima contexts, you’ll probably want to have something like this Fish function that I use:

```fish
# Defined in /Users/austin/.config/fish/functions/set-docker-host.fish @ line 1
function set-docker-host
    command -sq docker
    or return 1

    set -l host (docker context inspect --format '{{ .Endpoints.docker.Host }}')
    or return 1

    set -gx DOCKER_HOST $host
end
```

In `*sh`, this would be something like (untested):

```sh
# Defined in /Users/austin/.config/fish/functions/set-docker-host.fish @ line 1
function set-docker-host() {
  command -v docker > /dev/null 2>/dev/null || return 1

  host="$(docker context inspect --format '{{ .Endpoints.docker.Host }}')" || return 1

  export DOCKER_HOST="${host}"
}
```

**Reply by nipunsharma12** [reply]
Docker_host would need to be updated for the newer versions of colima
export DOCKER_HOST="unix://$HOME/.colima/default/docker.sock"

**Reply by halostatue** [reply]
Not if you’re using docker contexts the way that I suggested in my reply.

**Reply by nipunsharma12** [reply]
yes, you're right. 
-Thanks

**Comment by arun0009** [comment]
```sh
brew install docker-buildx
mkdir -p ~/.docker/cli-plugins
ln -sfn $HOMEBREW_PREFIX/opt/docker-buildx/bin/docker-buildx ~/.docker/cli-plugins/docker-buildx
```

**Comment by jasonwang178** [comment]
My solution:

```shell
brew install docker-buildx

mkdir -p ~/.docker/cli-plugins
ln -s $(which docker-buildx) ~/.docker/cli-plugins/docker-buildx

colima restart
docker buildx version
```

**Reply by 66Ton99** [reply]
One difference is that I used `nix-env -iA nixpkgs.docker-buildx`

**Comment by smiklos** [comment]
@abiosoft  Would it not make sense to include the lack of buildx and how to install it somewhere in the documentation? 

I'm using werf.io which by default uses the docker-engine. Building images didn't work until I disabled buildkit. I suppose installing buildx will also make it work but curios if any of the project should point this out for Colima users.

**Comment by debo** [comment]
Apologies for resuscitating this but my OCD is really killing me. I am running Colima and I am using `buildx` as well. What is really getting to me is that `docker context ls` works just fine but `docker buildx ls` reports an error on the default socket because it's point to docker default. Now, I know that if I set `DOCKER_HOST` to there right value the issue with `buildx` disappears but then `docker context ls` start throwing a warning because the env var is set.

Is there an alternative way that allow both commands to work just fine without needing to play around with env vars all the time? Many thanks for your help. Thank you.

**Comment by esafak** [comment]
You can install it with macports:
```
sudo port install docker-buildx-plugin
```
